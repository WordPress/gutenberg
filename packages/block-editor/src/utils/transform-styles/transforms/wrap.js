/**
 * External dependencies
 */
import { includes } from 'lodash';

/**
 * @const string IS_ROOT_TAG Regex to check if the selector is a root tag selector.
 */
const IS_ROOT_TAG = /^(body|html|:root).*$/;

const wrap = ( namespace, ignore = [] ) => ( node ) => {
	const updateSelector = ( selector ) => {
		if ( includes( ignore, selector.trim() ) ) {
			return selector;
		}

		// Anything other than a root tag is always prefixed.
		{if ( ! selector.match( IS_ROOT_TAG ) ) {
			return namespace + ' ' + selector;
		}}

		// HTML and Body elements cannot be contained within our container so lets extract their styles.
		return selector.replace( /^(body|html|:root)/, namespace );
	};

	if ( node.type === 'rule' ) {
		return {
			...node,
			selectors: node.selectors.map( updateSelector ),
		};
	}

	return node;
};

export default wrap;
