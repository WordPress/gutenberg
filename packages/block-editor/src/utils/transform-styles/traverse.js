/**
 * External dependencies
 */
import traverse from 'traverse';

/**
 * Internal dependencies
 */
import { parse, stringify } from './ast';

function traverseCSS( css, callback ) {
	try {
		const parsed = parse( css );

		const updated = traverse.map( parsed, function( node ) {
			if ( ! node ) {
				return node;
			}
			const updatedNode = callback( node );
			return this.update( updatedNode );
		} );

		return stringify( updated );
	} catch ( err ) {
		// eslint-disable-next-line no-console
		console.warn( 'Error while traversing the CSS: ' + err );

		return null;
	}
}

export default traverseCSS;
