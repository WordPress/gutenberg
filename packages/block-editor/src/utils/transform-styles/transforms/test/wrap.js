/**
 * Internal dependencies
 */
import traverse from '../../traverse';
import wrap from '../wrap';

describe( 'CSS selector wrap', () => {
	it( 'should wrap regular selectors', () => {
		const callback = wrap( '.my-namespace' );
		const input = `h1 { color: red; }`;
		const output = traverse( input, callback );

		expect( output ).toMatchSnapshot();
	} );

	it( 'should wrap multiple selectors', () => {
		const callback = wrap( '.my-namespace' );
		const input = `h1, h2 { color: red; }`;
		const output = traverse( input, callback );

		expect( output ).toMatchSnapshot();
	} );

	it( 'should ignore selectors', () => {
		const callback = wrap( '.my-namespace', 'body' );
		const input = `h1, body { color: red; }`;
		const output = traverse( input, callback );

		expect( output ).toMatchSnapshot();
	} );

	it( 'should replace root tags', () => {
		const callback = wrap( '.my-namespace' );
		const input = `body, h1 { color: red; }`;
		const output = traverse( input, callback );

		expect( output ).toMatchSnapshot();
	} );

	it( 'should ignore keyframes', () => {
		const callback = wrap( '.my-namespace' );
		const input = `
		@keyframes edit-post__fade-in-animation {
			from {
				opacity: 0;
			}
		}`;
		const output = traverse( input, callback );

		expect( output ).toMatchSnapshot();
	} );

	it( 'should ignore font-face selectors', () => {
		const callback = wrap( '.my-namespace' );
		const input = `
		@font-face {
			font-family: myFirstFont;
			src: url(sansation_light.woff);
		}`;
		const output = traverse( input, callback );

		expect( output ).toMatchSnapshot();
	} );

	it( 'should replace :root selectors', () => {
		const callback = wrap( '.my-namespace' );
		const input = `
		:root {
			--my-color: #ff0000;
		}`;
		const output = traverse( input, callback );

		expect( output ).toMatchSnapshot();
	} );
} );
