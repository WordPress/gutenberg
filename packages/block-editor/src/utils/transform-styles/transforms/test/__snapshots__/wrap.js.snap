// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CSS selector wrap should ignore font-face selectors 1`] = `
"@font-face {
font-family: myFirstFont;
src: url(sansation_light.woff);
}"
`;

exports[`CSS selector wrap should ignore keyframes 1`] = `
"@keyframes edit-post__fade-in-animation {
from {
opacity: 0;
}
}"
`;

exports[`CSS selector wrap should ignore selectors 1`] = `
".my-namespace h1,
body {
color: red;
}"
`;

exports[`CSS selector wrap should replace root tags 1`] = `
".my-namespace,
.my-namespace h1 {
color: red;
}"
`;

exports[`CSS selector wrap should wrap multiple selectors 1`] = `
".my-namespace h1,
.my-namespace h2 {
color: red;
}"
`;

exports[`CSS selector wrap should wrap regular selectors 1`] = `
".my-namespace h1 {
color: red;
}"
`;

exports[`CSS selector wrap should replace :root selectors 1`] = `
".my-namespace {
--my-color: #ff0000;
}"
`;
